### ISSUE TITLE
Add opional two-factor authentication for user accounts

### ISSUE THREAD

**Issue by grrrrr** [openstreetmap/openstreetmap-website#3476]
### Description

This was originally asked in #979 but was closed in 2015 due to lack of support.   I would like to re-address this given the time passed, coming up on 7 years, and hopefully more capacity is available.

It would be very useful to get [TOTP](https://en.wikipedia.org/wiki/Time-based_One-time_Password_Algorithm) based support on the website so that users and OSM volunteers can add an additional layer of protection to their accounts

**Comment by tomhughes** [comment]
I don't think that things have really changed very much since I wrote my comments on that bug - we would need to make substantial changes to other things before we consider this.

In reality it's probably only something we could do once we have finally managed to move the authentication to use a standard framework like devise which no doubt has the necessary tools available.

**Comment by grrrrr** [comment]
no problem Tom, thanks for getting back to me

**Comment by mmd-osm** [comment]
I have no idea whether the following blog post is of any use. It talks about TOTP in the context of Devise: https://reinteractive.com/articles/implementing-2FA-with-authenticator-apps-(totp)-in-rails

Some of the gems mentioned there appear to be still maintained: [devise_two_factor](https://github.com/tinfoil/devise-two-factor), [rotp](https://github.com/mdp/rotp) and [rqrcode](https://github.com/whomwah/rqrcode)

**Comment by mmd-osm** [comment]
By the way, most third party login providers should support 2FA. This might be a good alternative for the time being.

**Comment by pdschneider** [comment]
I would just like to add my two cents here.

I don't think I need to convince anyone about the importance of MFA for security in 2026 and it worries me a tiny bit that OSM still doesn't have it. The implementation sounds pretty tough by the looks of it, but this is the kind of thing I personally think should be a top priority. The last thing I would want is for my account to be compromised after putting so much time into contributing.

Is moving to Devise something that is being actively worked on?

**Comment by pablobm** [comment]
@pdschneider - I'm working on it. The main ticket is https://github.com/openstreetmap/openstreetmap-website/issues/6809. I just published a short update to let people know.
